src/pubsub/ua_pubsub_ns0.c

```diff
@@ -1849,6 +1849,149 @@ setSecurityKeysAction(UA_Server *server, const UA_NodeId *sessionId, void *sessi
+    /*Check whether the channel is encrypted according to specification*/
+    session_list_entry *session_entry;
+    LIST_FOREACH(session_entry, &server->sessions, pointers) {
+        if(UA_NodeId_equal(&session_entry->session.sessionId, sessionId)) {
+            if(session_entry->session.header.channel->securityMode !=
+               UA_MESSAGESECURITYMODE_SIGNANDENCRYPT)
+                return UA_STATUSCODE_BADSECURITYMODEINSUFFICIENT;
+        }
+    }
+
+    if(!server || !input)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    if(inputSize < 3)
+        return UA_STATUSCODE_BADARGUMENTSMISSING;
+    if(inputSize > 3 || outputSize > 5)
+        return UA_STATUSCODE_BADTOOMANYARGUMENTS;
+
+    /*check for types*/
+    if(!UA_Variant_hasScalarType(&input[0],
+                                 &UA_TYPES[UA_TYPES_STRING]) || /*SecurityGroupId*/
+       !UA_Variant_hasScalarType(&input[1],
+                                 &UA_TYPES[UA_TYPES_UINT32]) || /*StartingTokenId*/
+       !UA_Variant_hasScalarType(&input[2],
+                                 &UA_TYPES[UA_TYPES_UINT32])) /*RequestedKeyCount*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    /*check for types*/
+    if(input[0].type != &UA_TYPES[UA_TYPES_STRING] || /*SecurityGroupId*/
+       input[1].type != &UA_TYPES[UA_TYPES_UINT32] || /*StartingTokenId*/
+       input[2].type != &UA_TYPES[UA_TYPES_UINT32])  /*RequestedKeyCount*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_StatusCode retval = UA_STATUSCODE_BAD;
+    UA_UInt32 currentKeyCount = 1;
+    /* input */
+    UA_String *securityGroupId = (UA_String *)input[0].data;
+    UA_UInt32 startingTokenId = *(UA_UInt32 *)input[1].data;
+    UA_UInt32 requestedKeyCount = *(UA_UInt32 *)input[2].data;
+
+    UA_PubSubKeyStorage *ks =
+        UA_Server_findKeyStorage(server, *securityGroupId);
+    if(!ks)
+        return UA_STATUSCODE_BADNOTFOUND;
+
+    UA_Boolean executable = UA_FALSE;
+    UA_SecurityGroup *sg = UA_SecurityGroup_findSGbyName(server, *securityGroupId);
+    void *sgNodeCtx;
+    getNodeContext(server, sg->securityGroupNodeId, (void **)&sgNodeCtx);
+    executable = server->config.accessControl.getUserExecutableOnObject(
+        server, &server->config.accessControl, sessionId, sessionHandle, methodId,
+        methodContext, &sg->securityGroupNodeId, sgNodeCtx);
+
+    if(!executable)
+        return UA_STATUSCODE_BADUSERACCESSDENIED;
+
+    /* If the caller requests a number larger than the Security Key Service permits, then
+     * the SKS shall return the maximum it allows.*/
+    if(requestedKeyCount > sg->config.maxFutureKeyCount)
+        requestedKeyCount =(UA_UInt32) sg->keyStorage->keyListSize;
+    else
+        requestedKeyCount = requestedKeyCount + currentKeyCount; /* Add Current keyCount */
+
+    /*The current token is requested by passing 0.*/
+    UA_PubSubKeyListItem *startingItem = NULL;
+    if(startingTokenId == 0) {
+        startingItem = sg->keyStorage->currentItem;
```

Comment:
Can the currentItem be NULL?
If not, add an assert and document why this can never be the case.